### DISCUSSION TITLE
rbtree

### DISCUSSION THREAD

**Discussion by masatake** [util-linux/util-linux#1760] [Q&A] [answered]
@karelzak In lsfd, I need a dictionary mapping from pid to struct proc.

Writing a hash table is one of choice.
Introducing a rbtree implementation is another choice. 
Writing a rtbree is not easy for me.
So I'm thinking about copying https://github.com/forhappy/rbtree/blob/master/rbtree.h and .c to the util-linux's source tree. Is it acceptable?
The rbtree implementation is derived from Linux and I'm using it in Universal Ctags.

**Comment by karelzak** [comment, marked as answer]
Why do we need to search in the list of processes? For any filtering purpose? 

BTW, there is list_sort(), so it should be possible to use some binary search for the list of the proc structs -- but I have no clue how you want to use it (if you want to add new items to the directory after sort(), then it's useless).

misc-utils/hardlink.c uses tsearch() and twalk() for inode numbers. In this case, the binary search tree works well enough. Maybe we can use the same for lsfd rather than introduce another code. I guess the performance of the selected algorithm is rarely an issue when the application waste 95% of the time in read() :-)

Anyway, if all the possibilities are not good enough, we can add a new code; the ideal would be to add LGPL, BSD, or the public domain. GPL is also acceptable, although a little bit more problematic in the lib/ directory.

**Reply by masatake** [reply]
I'm thinking about printing comm for pid targeted by pidfd.
The dictionary is needed for getting comm from pid.

I didn't know tsearch(). I'm very surprised that such a useful function is available widely. tsearch() looks enough for the purpose. Thank you for your advice.
